### PR TITLE
Proposal for V1

### LINKED ISSUE #36: Hover color not working properly when cursor is right between icons
When rating, if I place the cursor right between, say, 4th and 5th stars, both will be empty while the first 3 are full, same for 3rd and 4th start with first 2 etc. It's a very small problem, but makes the rating look buggy. I'm using material design icons, but it also happens with the default icon. Looks like there's some sort of border around the icons, but it does not show when inspecting the elements. Any advice?

### LINKED ISSUE #38: Flicker issue on mousemove when getBoundingClientRect() returns non integer values.
https://github.com/dreyescat/react-rating/blob/8043bbcaa1c855f5d29b4ed7bb032703c9e20885/src/Rating.jsx#L158-L163

Since getBoundingClientRect()  can return sub-pixel values and clientX/Y are always rounded off, it causes flicker while mouse moves from one star to next for 1 pixel. I was able to fix this by using Math.floor on values returned by getBoundingClientRect().

### PR DESCRIPTION
**Pull request by mkarajohn**
I initially started this as another branch, before moving to this one, and my original goal was to clean up some code here and there, improve some interactions with the React way of doing things and break up concerns. Eventually I found myself making some major rewrites, removing parts that I felt made sense to be removed and renaming some things that cause the current API to change, leading to a major semver update if this gets merged.

## Breaking changes

* Changed `initialRate` to `initialValue`
* Removed the `placeHolderRate` and `placeholder` interfaces. I strongly believe that it only makes using the component more confusing, since the user can already set an initial value and I cannot think of any use case scenario that is not covered by `initialValue`.
* Changed `full` to `fullSymbol` to be more descriptive of what it is.
* Changed `empty` to `emptySymbol` to be more descriptive of what it is.
* Removed `onRate` callback function. `onChange` now serves a similar purpose.

## Overview of changes

* The `Rating` component is now broken down to two different components: `RatingContainer` and `Rating`. `RatingContainer` is the exported component
* `RatingContainer` is only concerned with holding the **actual value** of the current rating the user has assigned, i.e. the value the user is most likely interested in returning in a callback. For example, did the user click the 3rd symbol in a 5 symbol rating with the minimum value being 10 and the maximum value being 20, with a step of 2 with just 1 fraction? The `RatingContainer` will update its value with the value 16.
* `Rating` is now only concerned with managing what the user sees and interacts with. 
  * It holds the **display value**, i.e. what the user sees when they hover over the symbols.
  * It keeps track of whether the rating element has been clicked even once
  * It keeps track of whether the user is currently hovering over it
In the previous example, `Rating`'s internal display value would be updated to 6
* The simple explanation of how these 2 components interact with each other is the following: 
  * `RatingContainer` passes an appropriate display value to the `Rating`, based on the `RatingContainer`'s real value
  * `Rating` sets the display value that was passed down to it from props as an internal state. `Rating` renders based on the display value that is held in its state
  * All user interactions with the `Rating` element, except for clicking on it, only change its internal display value. The original display value that was passed to it is still held as a prop.
  * Clicking on the `Rating` element updates the `RatingContainer`'s value in its state, which in turn is passed down to `Rating` as a prop, which updates its internal state, which triggers a re-render.
* The above break up led to some code that mapped the previously used "index" to a value to be redundant and thus removed.
* In general I think code was simplified and I broke concerns as much as I could. I also renamed some functions and variables to be more descriptive.
* ~~Changed the way the active rating symbols are calculated and displayed. Previously, the symbols themselves where integral to calculations, and values passed by them were used in the computations, in order to calculate how many symbols to render as full. Now, the symbols themselves are not used at all as event targets and in calculations. How many symbols will be rendered as full and what value is passed in callbacks is now determined solely from the full width of the container, the position of the mouse cursor and the amount of symbols that exist within the container. [Relevant point in the code](https://github.com/dreyescat/react-rating/pull/49/files#diff-b6a4d22f9b15af7dfa4af864bbe2ded9R75)~~ See [this](https://github.com/dreyescat/react-rating/pull/49/commits/13883f4399d61cade46581d69b78cac2fbd54caa)
* ~~The above change has the downside that if for some reason the rating container is forced to break lines, it will stop working properly. This should not be happening however, since the rating container is supposed to always be in one line. If this turns to be an issue, I will try and think of a way to use the symbols without having the flickering. Worst case, we can revert to the old way of calculations.~~ See [this](https://github.com/dreyescat/react-rating/pull/49/commits/13883f4399d61cade46581d69b78cac2fbd54caa)
* Direction is now passed as a prop and is not calculated in any way from the components
* `onChange` now fires every time the internal state of the `Rating` gets changed but not `onClick`
* `onClick` returns the click event that happened ~~on the `Rating` container and not~~ on a specific symbol, ~~because we no longer add listeners to symbols~~ See [this](https://github.com/dreyescat/react-rating/pull/49/commits/13883f4399d61cade46581d69b78cac2fbd54caa)
* In general I tried to use every good practice I could think of to keep wasted renders to a minimum.
* Fixes #36 
* Fixes #38 

## Things that do not work

* The tests 😅. I am not used to them and I will probably need help if this is going to move forward. I manually tested it though next to a rating component created with the existing implementation, and the new implementation behaves the same.

### CODE REVIEW

**Comment by dreyescat** [comment]
Wow! Nice work!

I have just glossed over the list of changes. In general they look sound and consistent to me. I think it is a good moment to release a new major version with all your improvements.  Give me some time to have a deeper look and we can go for it.

By the way, would you mind to be a collaborator? I really think that after this rewrite you are perfectly capable to be a *core* contributor :wink:.

**Comment by mkarajohn** [comment]
I'd be glad to be a collaborator!

Also, I would like to know what you think the callback API should be? Currently, here's a detailed description of how `onChange` and `onClick` work (note: `state` and `props` are in regards to `Rating`):

* `onChange`: Only fires when there is change in the `state.displayValue` of the `Rating` component and returns the current real value (the `value` that `RatingContainer` holds in its state). The `state.displayValue` is how the `Rating` component keeps track over which symbol (or part of it) you are currently hovering your mouse over. A change in the `state.displayValue` can happen in the following ways:
  * The user hovers over the symbols of the `Rating` component. Note that if a `Rating` is already set to a value and you mouse enter and hover over that value, there is no change happening in `state.displayValue` so the `onChange` callback does not fire. 
 * The `RatingContainer` passes down a new `props.establishedValue` to `Rating`, which the latter will use to update its `state.displayValue`, so that `state.displayValue = props.establishedValue`. If the `props.establishedValue` is changed because the user clicked on a symbol, the `onChange` will not fire, because the sequence of events is that the user first hovered over the symbol, which set the `state.displayValue` to a new value, triggering `onChange`, and then clicked on it, which caused the `props.establishedValue` to change, but because it is the same as the `state.displayValue` no  changes happened, so `onChange` does not fire. 
* `onClick` only fires when the user clicks somewhere inside the `Rating` component and returns the real value that corresponds to where the user clicked as well as the `click` event. 

I am contemplating if it would make sense to add callbacks for entering and leaving individual symbols, and what values they should return. Keep in mind a symbol can actually represent many values if `fractions > 1`. What do you think, what would be a complete callback API in your opinion?

**Comment by mkarajohn** [comment]
I managed to find a way to use the more correct way of value calculation, based on the position of the cursor over specific symbols while also not having the annoying flickering 🎉

Updated all the relevant parts from above.

**Comment by dreyescat** [comment]
I like your idea of reducing the callbacks to two. Originally there were only two callbacks:
- `onChange`: This was supposed to be your `onClick`. Only fired when the established value (not the display value) changed. That happened to be when a symbol was clicked on a different value than current one.
- `onRate`: This was supposed to be your `onChange`. Only fired when the displayed value changed. Sort of `onMouseEnter`/`onMouseLeave` with the *not so elegant* `undefined` value on leave.
Then the third callback `onClick` was introduced in #9. In this case the original `onChange` was useless but I didn't want to break compatibility so the `onChange` - `onClick` duality was born.

Now that we are going to release a major version I prefer your proposal. `onChange` for display changes and `onClick` for click (and actual value change). Right now I can't think of a scenario where the established value could change without clicking a symbol (until someone somewhere comes up with some new usage we haven't imagined yet :wink:).

**Comment by djkwagala** [comment]
When you say `onChange for display changes and onClick for click (and actual value change).` its confusing. I thought: 

- `onChange`  will fire when an actual value change has occured. 
- `onClick` would fire for any click on a symbol whether there's an actual rating change or not.

**Comment by dreyescat** [comment]
@djkwagala Maybe it is a little bit confusing because we are reusing the same callback name as in `v0` for a different goal. In `v1` the `onChange` callback sort of replaces the original `onRate`. While the original `onChange` is just removed.

**Comment by djkwagala** [comment]
Its clear.

**Comment by dreyescat** [comment]
@djkwagala comment made me think... If `onClick` would fire for any click regardless the established value has been changed or not, are we going to delegate to the user to know if this click was an actual change?

**Comment by dreyescat** [comment]
> I am also contemplating if it would make sense to add callbacks for entering and leaving individual symbols, and what values they should return. Keep in mind a symbol can actually represent many values if `fractions > 1` or `step > 1`. What would be a complete callback API in your opinion? 

Uhm. Here I am not sure I would add callbacks at symbol level. So far I have not received any request with this level of granularity. I would expect to work at value level. I mean, I expect to be informed when I am over a value not a symbol. I don't see a practical use that cannot be solved with `onChange`. Anyway I am not closed to add them. Maybe more creative people can find it useful 😃 .

**Comment by dreyescat** [comment]
> * Removed the `placeHolderRate` and `placeholder` interfaces. I strongly believe that it only makes using the component more confusing, since the user can already set an initial value and I cannot think of any use case scenario that is not covered by `initialValue`.

I understand your feelings here. I was of your opinion too. But #6 made me think it twice. Having a `placeholder` might be useful for *three-state* symbols. How could we simulate those Netflix like ratings?

**Comment by dreyescat** [comment]
> * The tests 😅. I am not used to them and I will probably need help if this is going to move forward. I manually tested it though next to a rating component created with the existing implementation, and the new implementation behaves the same.

Don't worry! I will take care of tests.

**Comment by mkarajohn** [comment]
@dreyescat @djkwagala 

> When you say onChange for display changes and onClick for click (and actual value change). its confusing. I thought:

> - onChange will fire when an actual value change has occured.
> - onClick would fire for any click on a symbol whether there's an actual rating change or not.

`onClick` will fire regardless if you clicked on the same value or not.

`onChange` will fire when a "perceived" value changes. By perceived value, I mean the value that changes when the user hovers over the `Rating` element. Would it help if I renamed `displayValue` to `perceivedValue`? Also note that `onChange` will fire even if `quiet === true` and there is no actual visual indication to the user. `onChange` will not fire if you keep hovering over the same value, or if, for example the `Rating` had an initial rating of 3 and you mouse enter over the 3.

> If onClick would fire for any click regardless the established value has been changed or not, are we going to delegate to the user to figure out if this click was an actual change?

Yes. `onClick` implies triggering on the click event, regardless. You clicked somewhere? You get the callback to fire, regardless if you clicked the same thing.

In general, this 2 callback API, has a callback that guarantees firing on a perceived value change, but does not guarantee that this change came from the user's explicit intent (i.e. clicking on a symbol), and a callback that guarantees firing on any click but does not guarantee that it only fires when the click caused a change in value.

Would it make sense to have a callback option that fires only when the user clicks somewhere and the click actually caused a value change? Pretty much a callback that takes from the user the burden of checking if each click actually changed the value. It could be called `onSet` or something. I think it would make sense. The other 2 callbacks would still fire as originally intended if used in tandem. 

We could also add an `onMouseEnter` and `onMouseLeave`. For example use `onMouseEnter` to add a DOM element which then `onChange` would use to display the current perceived value and `onMouseLeave` would remove that DOM element

>Uhm. Here I am not sure I would add callbacks at symbol level.

Cool, lets leave it as is then for now.

> How could we simulate those Netflix like ratings?

My issue with that was that it needde the `Rating` component to have an `undefined` value in order for the placeholder symbols to show up, or else, if an `initialRate` was used, even if it was 0, it would override the placeholder symbols. I don't think the `Rating` should ever have `undefined` as value, I think it should always start somewhere. 

One way I can think for it so it makes sense would be to use like this:

* If you set both `initialRate` and `placeholderRate` the latter is ignored (it's how it works now)
* If you just use `placeholderRate` internally it will set the actual value to be the same as the placeholder rating value, so we do not have `undefined` as an internal value. 
* Document this behaviour in the docs 😄

**Comment by mkarajohn** [comment]
...Coming to think of it, maybe it makes sense to initialize the Rating with no value at all. Usually this is how ratings start, I think...

**Comment by dreyescat** [comment]
Yep. I was going to suggest the same... A no value in this case is a *valid* value and having it by default is the most common use case.

> We could also add an `onMouseEnter` and `onMouseLeave` (on the `Rating` container level).

You mean on entering/leaving a rating value? If so, what do you think if we get rid of the `onChange` and replace it with the `onMouseEnter` and provide the opposite `onMouseLeave` (for consistency)? My feeling is that people is going to get confused with an `onChange` event that tracks the `displayedValue`/`perceivedValue` (not sure which name is better...) instead of the actual value. Try to imagine explaining someone that a `Rating` in `quiet` mode is firing `onChange` events. From user point of view nothing *changed*.

> Would it make sense to have a callback option that fires only when the user clicks somewhere and the click actually caused a value change? Pretty much a callback that takes from the user the burden of checking if each click actually changed the value. It could be called `onSet` or something. I think it would make sense. The other 2 callbacks would still fire as originally intended if used in tandem.

Ups, my subconcious made me read `onChange` instead of `onSet` here :wink:.

> My issue with that was that it needde the `Rating` component to have an `undefined` value in order for the placeholder symbols to show up, or else, if an `initialRate` was used, even if it was 0, it would override the placeholder symbols. I don't think the `Rating` should ever have `undefined` as value, I think it should always start somewhere.

Did you change your mind now that having `undefined` value makes sense?

I will play a little bit during the weekend to familiarize with the new arrangement so I can give you more feedback. So far I have only read through the code (really good, by the way). I am keen to see it working on my own laptop 😄.

You've done a great job! :clap:.

**Comment by mkarajohn** [comment]
>Try to imagine explaining someone that a Rating in quiet mode is firing onChange events.

Yeah, that kind of bugged me too, but I kept it like this to keep it somewhat consistent with how it works now, where you can hover over the symbols in quiet mode and have the value being displayed next to it. Are you suggesting to drop this possibility?

>Did you change your mind now that having undefined value makes sense?

Yep!

**Comment by deeartrama88** [comment]
Hey guys, i'm new here, just wanna ask question about version, I'm instaling react-rating via npm, version is 0.7.0, all is fine, except that i still see warging about prop-types.. i see you had fixed this, but it's not released yet, rigth?

**Comment by dreyescat** [comment]
@deeartrama88 Release [v0.7.1](https://github.com/dreyescat/react-rating/releases/tag/v0.7.1) includes the commit that fixes this warning.

**Comment by dreyescat** [comment]
> Yeah, that kind of bugged me too, but I kept it like this to keep it somewhat consistent with how it works now, where you can hover over the symbols in quiet mode and have the value being displayed next to it. Are you suggesting to drop this functionality?

Nope. I am not suggesting removing this functionality. I also think it makes it consistent to have a callback that is called when you **hover over** the symbols (or more precisely, over rating values) even in quiet mode. For me it is a matter of names :wink:. Instead of reusing `onChange`, that can be misleading, use another name `onHover`, `onEnter`, or whatever name that avoids confusion. Maybe [the web events reference](https://developer.mozilla.org/en-US/docs/Web/Events) could be helpful. I like to see the rating as an specialized `<select>` element.

**Comment by deeartrama88** [comment]
@dreyescat thanks)

**Comment by dreyescat** [comment]
@mkarajohn I don't know how to merge your pull request into another branch than `master`. Could you please use `v1` instead? I just created this new branch. Or, if you feel comfortable, just create the branch yourself and push it (you should have the right permissions now).

File: src/Rating.jsx [hunk 1]
```diff
@@ -2,238 +2,174 @@
...
+    this.symbolClick = this.symbolClick.bind(this);
   }
 
   componentWillReceiveProps(nextProps) {
-    const rate = nextProps.initialRate !== undefined ?
-      nextProps.initialRate : nextProps.placeholderRate;
     this.setState({
-      index: indexOf(nextProps, rate),
-      selected: nextProps.initialRate !== undefined
+      dirty: ((this.props.establishedValue !== nextProps.establishedValue) && !this.state.dirty) ? true : this.state.dirty,
```
Comment (dreyescat) [inline src/Rating.jsx]: What is `dirty` used for?

**Comment by mkarajohn** [comment]
@dreyescat 

> Could you please use v1 instead? I

Hm, I cannot push to this v1, it says I do not have permissions. I will try changing the base of this PR from master to v1

**Comment by mkarajohn** [comment]
@dreyescat I made an update to the method API and to the README. Please check it out (the readme mostly) and tell me if it is more in line with what you had in mind. I think it is close to (if not already) making sense.

**Comment by dreyescat** [comment]
I have merged it to `v1` branch. Now we can work together in this branch. I am going to create an issue to track/discuss decisions and TODO's.

@mkarajohn Did you receive my invitation to become a collaborator? Being a collaborator should give you full access to the repo, or at least this is what I thought...

**Comment by mkarajohn** [comment]
@dreyescat ah, yes, I just did, I finally thought checking my notifications here as well, in addition to my email which had not received anything

**Comment by dreyescat** [comment]
Perfect! I have created #50. We can track there what we need to release `v1`.

**Pull request merged**